### PR TITLE
add to UdpPacket a UdpPayloadCustomDecoderFunc static function

### PR DESCRIPTION
**Pull request by jgaulon**
Please find in the PR the addition into UdpPacket of a callback function for payload decoding.
Regards, Julien

### CODE REVIEW

File: PacketDotNet/UdpPacket.cs [hunk 1]
```diff
@@ -25,6 +25,14 @@ namespace PacketDotNet
     /// </summary>
     public sealed class UdpPacket : TransportPacket
     {
+        /// <summary>
+        /// Callback function for UdpPacket Payload decoding.
+        /// First parameter is the payload, second parameter if the UdpPacket itself
+        /// returned value is the decoded payload as Packet or null if payload is invalid or not supported
+        /// then internal decoding will continue
+        /// </summary>
+        public static Func<ByteArraySegment, UdpPacket, Packet> UdpPayloadCustomDecoderFunc;
```
Comment (PhyxionNL) [inline PacketDotNet/UdpPacket.cs:34]: I think we can move this to TransportPacket (and TransportPacket as second parameter) so TcpPacket can also use it. The result should be a PacketOrByteArraySegment to also make it possible to return a byte array segment if it's needed. 
The name can be CustomPayloadDecoder.

Comment (jgaulon) [reply PacketDotNet/UdpPacket.cs:34]: Sure I can, but this would contradict a little bit the stuff done packet-connection isn't it?

Comment (PhyxionNL) [reply PacketDotNet/UdpPacket.cs:34]: It's a separate project and this would also be useful for Tcp payload parsing 🙂

**Comment by jgaulon** [comment]
@PhyxionNL is it ok now?

File: PacketDotNet/TransportPacket.cs [hunk 2]
```diff
@@ -20,6 +21,14 @@ namespace PacketDotNet
     /// </summary>
     public abstract class TransportPacket : Packet
     {
+        /// <summary>
+        /// Callback function for UdpPacket Payload decoding.
+        /// First parameter is the payload, second parameter if the UdpPacket itself
+        /// returned value is the decoded payload as Packet or null if payload is invalid or not supported
+        /// then internal decoding will continue
+        /// </summary>
```
Comment (PhyxionNL) [inline PacketDotNet/TransportPacket.cs:29]: This comment should be updated to reflect transport packet.

File: PacketDotNet/TransportPacket.cs [hunk 3]
```diff
@@ -20,6 +21,14 @@ namespace PacketDotNet
...
+        public static Func<ByteArraySegment, TransportPacket, Packet> CustomPayloadDecoder;
```
Comment (PhyxionNL) [inline PacketDotNet/TransportPacket.cs:30]: Can you change the result to a ```PacketOrByteArraySegment```? Then we'd also be able to close #103 as well as it'd be possible to return the payload itself.

**Review by PhyxionNL — CHANGES_REQUESTED** [review]
Few misc, rest looks good 🙂

File: PacketDotNet/TransportPacket.cs (hunk 2 above)
Comment (jgaulon) [reply PacketDotNet/TransportPacket.cs:29]: oh, sorry, yes of course.

Comment (PhyxionNL) [reply PacketDotNet/TransportPacket.cs:29]: No problem 👍

File: PacketDotNet/TransportPacket.cs (hunk 3 above)
Comment (jgaulon) [reply PacketDotNet/TransportPacket.cs]: done

File: PacketDotNet/TransportPacket.cs (hunk 2 above)
Comment (jgaulon) [reply PacketDotNet/TransportPacket.cs]: done

**Pull request merged**